File: Packages/LineScopeCore/Sources/LineScopeCore/PixelBuffer.swift
import CoreGraphics
import Foundation

/// A color sample with straight (non-premultiplied) alpha, sRGB-encoded components in 0...1.
public struct RGBA: Sendable, Equatable {
    public var r: Float
    public var g: Float
    public var b: Float
    public var a: Float

    public init(r: Float, g: Float, b: Float, a: Float = 1) {
        self.r = r
        self.g = g
        self.b = b
        self.a = a
    }
}

/// Interleaved RGBA Float32 image. Row 0 is the top row.
/// Components are sRGB-encoded (not linearized) with straight alpha, so values match what is displayed.
public struct PixelBuffer: Sendable, Equatable {
    public let width: Int
    public let height: Int
    public var data: [Float]

    public init(width: Int, height: Int, data: [Float]) {
        precondition(width > 0 && height > 0 && data.count == width * height * 4)
        self.width = width
        self.height = height
        self.data = data
    }

    public init(width: Int, height: Int, fill: RGBA = RGBA(r: 0, g: 0, b: 0, a: 1)) {
        var data = [Float](repeating: 0, count: width * height * 4)
        for i in stride(from: 0, to: data.count, by: 4) {
            data[i] = fill.r
            data[i + 1] = fill.g
            data[i + 2] = fill.b
            data[i + 3] = fill.a
        }
        self.init(width: width, height: height, data: data)
    }

    public subscript(x: Int, y: Int) -> RGBA {
        get {
            let i = (y * width + x) * 4
            return RGBA(r: data[i], g: data[i + 1], b: data[i + 2], a: data[i + 3])
        }
        set {
            let i = (y * width + x) * 4
            data[i] = newValue.r
            data[i + 1] = newValue.g
            data[i + 2] = newValue.b
            data[i + 3] = newValue.a
        }
    }

    // MARK: Premultiplication (used by resampling and Core Image, which expect premultiplied data)

    func premultiplied() -> [Float] {
        var out = data
        for i in stride(from: 0, to: out.count, by: 4) {
            let a = out[i + 3]
            out[i] *= a
            out[i + 1] *= a
            out[i + 2] *= a
        }
        return out
    }

    static func fromPremultiplied(width: Int, height: Int, _ p: [Float]) -> PixelBuffer {
        var out = p
        for i in stride(from: 0, to: out.count, by: 4) {
            let a = min(max(out[i + 3], 0), 1)
            out[i + 3] = a
            if a > 1e-6 {
                out[i] = min(max(out[i] / a, 0), 1)
                out[i + 1] = min(max(out[i + 1] / a, 0), 1)
                out[i + 2] = min(max(out[i + 2] / a, 0), 1)
            } else {
                out[i] = 0
                out[i + 1] = 0
                out[i + 2] = 0
            }
        }
        return PixelBuffer(width: width, height: height, data: out)
    }
}

// MARK: - Core Graphics bridging

extension PixelBuffer {
    static let sRGB = CGColorSpace(name: CGColorSpace.sRGB)!

    /// Decodes any CGImage into 8-bit sRGB and converts it to a float buffer.
    public init?(cgImage: CGImage) {
        let w = cgImage.width
        let h = cgImage.height
        guard w > 0, h > 0 else { return nil }
        var bytes = [UInt8](repeating: 0, count: w * h * 4)
        let drawn = bytes.withUnsafeMutableBytes { raw -> Bool in
            guard let ctx = CGContext(
                data: raw.baseAddress, width: w, height: h, bitsPerComponent: 8, bytesPerRow: w * 4,
                space: PixelBuffer.sRGB, bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
            ) else { return false }
            ctx.interpolationQuality = .none
            ctx.draw(cgImage, in: CGRect(x: 0, y: 0, width: w, height: h))
            return true
        }
        guard drawn else { return nil }
        var p = [Float](repeating: 0, count: w * h * 4)
        for i in 0..<p.count { p[i] = Float(bytes[i]) / 255 }
        self = PixelBuffer.fromPremultiplied(width: w, height: h, p)
    }

    /// Renders the buffer as an 8-bit premultiplied sRGB CGImage for display or export.
    public func makeCGImage() -> CGImage? {
        var bytes = [UInt8](repeating: 0, count: data.count)
        for i in stride(from: 0, to: data.count, by: 4) {
            let a = min(max(data[i + 3], 0), 1)
            bytes[i] = UInt8((min(max(data[i], 0), 1) * a * 255).rounded())
            bytes[i + 1] = UInt8((min(max(data[i + 1], 0), 1) * a * 255).rounded())
            bytes[i + 2] = UInt8((min(max(data[i + 2], 0), 1) * a * 255).rounded())
            bytes[i + 3] = UInt8((a * 255).rounded())
        }
        guard let provider = CGDataProvider(data: Data(bytes) as CFData) else { return nil }
        return CGImage(
            width: width, height: height, bitsPerComponent: 8, bitsPerPixel: 32, bytesPerRow: width * 4,
            space: PixelBuffer.sRGB, bitmapInfo: CGBitmapInfo(rawValue: CGImageAlphaInfo.premultipliedLast.rawValue),
            provider: provider, decode: nil, shouldInterpolate: false, intent: .defaultIntent
        )
    }
}

File: Windows/LineScope.Core/PixelBuffer.cs
namespace LineScope.Core;

/// <summary>A color sample with straight (non-premultiplied) alpha, sRGB-encoded components in 0…1.</summary>
public readonly record struct Rgba(float R, float G, float B, float A = 1);

/// <summary>
/// Interleaved RGBA Float32 image. Row 0 is the top row.
/// Components are sRGB-encoded (not linearized) with straight alpha, so values match what is displayed.
/// Treat instances as values: operations return new buffers instead of mutating their input.
/// </summary>
public sealed class PixelBuffer : IEquatable<PixelBuffer>
{
    public int Width { get; }
    public int Height { get; }
    public float[] Data { get; }

    public PixelBuffer(int width, int height, float[] data)
    {
        if (width <= 0 || height <= 0 || data.Length != width * height * 4)
            throw new ArgumentException("Invalid pixel buffer dimensions.");
        Width = width;
        Height = height;
        Data = data;
    }

    public PixelBuffer(int width, int height) : this(width, height, new Rgba(0, 0, 0, 1)) { }

    public PixelBuffer(int width, int height, Rgba fill)
        : this(width, height, new float[Math.Max(width, 0) * Math.Max(height, 0) * 4])
    {
        for (int i = 0; i < Data.Length; i += 4)
        {
            Data[i] = fill.R;
            Data[i + 1] = fill.G;
            Data[i + 2] = fill.B;
            Data[i + 3] = fill.A;
        }
    }

    public Rgba this[int x, int y]
    {
        get
        {
            int i = (y * Width + x) * 4;
            return new Rgba(Data[i], Data[i + 1], Data[i + 2], Data[i + 3]);
        }
        set
        {
            int i = (y * Width + x) * 4;
            Data[i] = value.R;
            Data[i + 1] = value.G;
            Data[i + 2] = value.B;
            Data[i + 3] = value.A;
        }
    }

    public PixelBuffer Clone() => new(Width, Height, (float[])Data.Clone());

    // Premultiplication (used by resampling and the blur filters, which work on premultiplied data)

    internal float[] Premultiplied()
    {
        var output = (float[])Data.Clone();
        for (int i = 0; i < output.Length; i += 4)
        {
            float a = output[i + 3];
            output[i] *= a;
            output[i + 1] *= a;
            output[i + 2] *= a;
        }
        return output;
    }

    /// <summary>Un-premultiplies (in place) and clamps to 0…1.</summary>
    internal static PixelBuffer FromPremultiplied(int width, int height, float[] p)
    {
        for (int i = 0; i < p.Length; i += 4)
        {
            float a = Math.Clamp(p[i + 3], 0, 1);
            p[i + 3] = a;
            if (a > 1e-6f)
            {
                p[i] = Math.Clamp(p[i] / a, 0, 1);
                p[i + 1] = Math.Clamp(p[i + 1] / a, 0, 1);
                p[i + 2] = Math.Clamp(p[i + 2] / a, 0, 1);
            }
            else
            {
                p[i] = 0;
                p[i + 1] = 0;
                p[i + 2] = 0;
            }
        }
        return new PixelBuffer(width, height, p);
    }

    // 8-bit bridging (the Windows counterpart of the Core Graphics bridge). The byte layout is
    // premultiplied BGRA, which is what WIC's and WPF's Pbgra32 pixel format expects.

    /// <summary>Converts decoded 8-bit premultiplied BGRA pixels to a float buffer.</summary>
    public static PixelBuffer FromPremultipliedBgra8(int width, int height, byte[] bytes)
    {
        if (bytes.Length < width * height * 4) throw new ArgumentException("Too few bytes.");
        var p = new float[width * height * 4];
        for (int i = 0; i < p.Length; i += 4)
        {
            p[i] = bytes[i + 2] / 255f;
            p[i + 1] = bytes[i + 1] / 255f;
            p[i + 2] = bytes[i] / 255f;
            p[i + 3] = bytes[i + 3] / 255f;
        }
        return FromPremultiplied(width, height, p);
    }

    /// <summary>Renders the buffer as 8-bit premultiplied BGRA for display or export.</summary>
    public byte[] ToPremultipliedBgra8()
    {
        var bytes = new byte[Data.Length];
        for (int i = 0; i < Data.Length; i += 4)
        {
            float a = Math.Clamp(Data[i + 3], 0, 1);
            bytes[i + 2] = ToByte(Math.Clamp(Data[i], 0, 1) * a);
            bytes[i + 1] = ToByte(Math.Clamp(Data[i + 1], 0, 1) * a);
            bytes[i] = ToByte(Math.Clamp(Data[i + 2], 0, 1) * a);
            bytes[i + 3] = ToByte(a);
        }
        return bytes;
    }

    private static byte ToByte(float v) => (byte)MathF.Round(v * 255, MidpointRounding.AwayFromZero);

    public bool Equals(PixelBuffer? other) =>
        other is not null && Width == other.Width && Height == other.Height && Data.AsSpan().SequenceEqual(other.Data);

    public override bool Equals(object? obj) => Equals(obj as PixelBuffer);

    public override int GetHashCode() => HashCode.Combine(Width, Height, Data.Length);
}
